### PR TITLE
Issue #125 PCTCompile : ignore changes in defined includes.

### PR DESCRIPTION
**Pull request by axelebnerpcsat**
new Parameter: ignoreIncludes.

Hi Gilles,

that are my changes, please check it.
What is to do to add that new parameter to PCTCompileExt?
Get the error: PCTCompileExt doesn't support the "ignoreIncludes" attribute.

Axel

### CODE REVIEW

File: src/progress/pct/v10/pctCompile.p [hunk 1]
```diff
@@ -394,6 +400,12 @@ FUNCTION CheckIncludes RETURNS LOGICAL (INPUT f AS CHARACTER, INPUT TS AS DATETI
             ASSIGN TimeStamps.ttFile = IncFile
                    TimeStamps.ttFullPath = SEARCH(IncFile).
             ASSIGN TimeStamps.ttMod = getTimeStampF(TimeStamps.ttFullPath).
+            IF lignoreIncludes AND CAN-DO(cignoreIncludes, IncFile) THEN /* fake older date, to be not relevant */
```
Comment (cverbiest) [inline src/progress/pct/v10/pctCompile.p]: tiny remark 
I don't now if getting a timestamp takes a long time but I wouldn't execute it if I'm about to ignore the result. I'd move the getTimeStampF to the else of "IF lignoreIncludes AND CAN-DO".

On the other hand it might be better if you add a flag ignore to the TimeStamps table. It could be easier for debugging purposes to see that a timestamp is ignored than to see a fake timestamp.

just my 0,02€

**Comment by cverbiest** [comment]
I like the idea but I got confused by the naming. 
ignoreIncludes sounds like a logical property to me, personally I'd prefer ignoreIncludeList or ignoreIncludeCando .

**Comment by jakejustus** [comment]
Hi,
Thanks for this PR. I'll have a look as soon as possible. Did you create unit test for this attribute? If so, could you include them as part of this PR?

**Comment by jakejustus** [comment]
Would ignoredIncludes be better as the attribute name ?

**Comment by cverbiest** [comment]
yes, ignoredIncludes does not give the impression that it's a logical

File: src/progress/pct/v10/pctCompile.p (hunk 1 above)
Comment (axelebnerpcsat) [inline src/progress/pct/v10/pctCompile.p]: You are right, a field in the TimeStamps is smarter.

**Comment by axelebnerpcsat** [comment]
Thank you for your feedback. My next steps are:
1. add the unittest. Are there any rules?
2. naming of the parameter, how about: ExceptIncludeList, IcludesToExceptList,...? Please give me a name. In that case I'm emotionless.
3. Logic: add a field in the TEMP-TABLE TimeStamps named lToExcept as LOGICAL
4. Performance: I would prefer to let it that way, because the Logical lignoreIncludes is TRUE when the parameter was set. After that compare the filter with the actual file.
`IF lignoreIncludes AND CAN-DO(cignoreIncludes, IncFile) THEN ...`
5. add feature to PCTCompileExt: please give me a clue.
6. run all unittests: how to? `ant -f tests.xml test` ?

**Comment by jakejustus** [comment]
1. Just use tests/PCTCompile/testXX as a template, and test as many edge cases as possible. Test logic is in src/java/test/com/phenix/pct/PCTCompileTest.java. Don't forget annotations.
2. Would use ignoredIncludes
3. Yes
4. No opinion for now, I'll have a closer look at the code when back from my trip. Do what you think is the best :)
5. This can be safely skipped for now.
6. Yes. You can also change the <ClassFileSet> node just to include the PCTCompileTest (to speed up local tests)

**Comment by axelebnerpcsat** [comment]
Ready for review.

**Comment by jakejustus** [comment]
I only received a notification for the initial commit, not for the following ones, reason why I didn't do anything yet. I'll have a look this afternoon.

**Comment by jakejustus** [comment]
From a fresh build, I'm getting the following errors :
- Build error on PCTCompileTest.java line 846 (mod1 is already defined)
- And a test failure in your unit test :

com.phenix.pct.PCTCompileTest:test50
Test Messages
Created dir: C:\Users\gquer_000\Documents\pct-pr\testbox\PCTCompile\test50\build
PCTCompile - Progress Code Compiler
2 file(s) compiled
PCTCompile - Progress Code Compiler
ignoring changes in: src\test1.i
0 file(s) compiled
PCTCompile - Progress Code Compiler
ignoring changes in: src\test1.i
0 file(s) compiled
Failure

expected:<true> but was:<false>
org.testng.Assert.fail(Assert.java:89)
at org.testng.Assert.failNotEquals(Assert.java:489)
at org.testng.Assert.assertTrue(Assert.java:37)
at org.testng.Assert.assertTrue(Assert.java:47)
at com.phenix.pct.PCTCompileTest.test50(PCTCompileTest.java:856)
24 lines not shown

**Comment by jakejustus** [comment]
With <sleep> before the <touch> statements, the test case is OK. Timestamps for filesystems are usually going up to 1 or 2 seconds, so if everything is done is less than 1 second, then the test case can fail.

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,9 @@
 /classDocTemp/
 /profiler/
 src/java/com/phenix/pct/PCT.properties
+/tests/**/*.r
```
Comment (jakejustus) [inline .gitignore]: Please remove from pull request, I won't merge .gitignore

File: pct.build.properties [hunk 3]
```diff
@@ -1,5 +1,5 @@
 DLC8=\\
 DLC9=\\
-DLC10=\\
+DLC10=d:\\dlc102b
```
Comment (jakejustus) [inline pct.build.properties]: Please remove from pull request, I won't merge this file

File: src/test/com/phenix/pct/PCTCompileTest.java [hunk 4]
```diff
@@ -827,4 +827,46 @@ public void test49() {
...
+        long mod1 = f1.lastModified();
+        File f2 = new File("PCTCompile/test50/build/test2.r");
+        assertTrue(f2.exists());
+        long mod2 = f2.lastModified();
+		
+        executeTarget("test2"); /* nothing should compile */
+        File f3 = new File("PCTCompile/test50/build/test.r");
+        assertTrue(f3.exists());
+		assertFalse(f3.lastModified() > mod1);
+        long mod1 = f1.lastModified();
```
Comment (jakejustus) [inline src/test/com/phenix/pct/PCTCompileTest.java]: Doesn't compile

File: tests/PCTCompile/test50/build.xml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    <mkdir dir="build" />
+    <PCTCompile destDir="build" dlcHome="${DLC}" relativePaths="true">
+      <fileset dir="src" includes="*.p" />
+      <propath location="src" />
+      <Profiler enabled="${PROFILER}" coverage="true" outputDir="profiler" />
+    </PCTCompile>
+  </target>
+
+  <target name="test2">
+    <touch file="src\test1.i" />
```
Comment (jakejustus) [inline tests/PCTCompile/test50/build.xml]: Add &lt;sleep seconds="2" /&gt; before each &lt;touch /&gt; node

File: tests/PCTCompile/test50/src/test.p [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{src\test2.i}
```
Comment (jakejustus) [inline tests/PCTCompile/test50/src/test.p]: Broken test on Unix, please use forward slash instead of backslash

File: tests/PCTCompile/test50/src/test2.p [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{src\test3.i}
```
Comment (jakejustus) [inline tests/PCTCompile/test50/src/test2.p]: Broken test on Unix, please use forward slash instead of backslash

**Comment by axelebnerpcsat** [comment]
Changed the slashes. Good to hear that it works. Would be fine when I also could test it.
Maybe you have a clue. I don't want to push so stupid bugs.

File: tests/PCTCompile/test50/src/test2.i [hunk 8]
```diff
@@ -0,0 +1 @@
+DEFINE VARIABLE cTest AS CHARACTER INIT {src\test1.i} NO-UNDO.
```
Comment (jakejustus) [inline tests/PCTCompile/test50/src/test2.i]: Still a backslash here, didn't see it at the first time, but unit tests failed again !

**Comment by jakejustus** [comment]
And in order to execute tests, I'm always opening a proenv (with 11.6 in my case), then cd to the base directory of pct, followed by `ant -file tests.xml test`

**Comment by jakejustus** [comment]
Perfect, green light for unit tests. Just a last request before merging the PR: can you remove your additional testEA section from tests.xml, that's not directly related to the additional feature.

**Comment by axelebnerpcsat** [comment]
Now with resetted tests.xml.

**Comment by axelebnerpcsat** [comment]
Thank you for your help and merging into PR. What is to do to have the feature in the PCTCompileExt, I need it also there for numThreads.

**Comment by jakejustus** [comment]
Hi,
This comment to let you know that PCTCompile and PCTCompileExt are now merged, so your attribute ignoredIncludes will be available from a multi-threaded build. If you can try the pre-release and find any defect, that would be perfect :)

**Comment by axelebnerpcsat** [comment]
Hi,

at the moment we only use OE10. "pct/v10/pctCompile.p" is missing.
Please add the version for OE10, then I can try it again.

Thank you
Axel

[bootstrapCompile] PCTCompile - Progress Code Compiler
[bootstrapCompile] *\* "pct/v10/pctCompile.p" nicht gefunden. (293)

BUILD FAILED
d:\pcsdev\pct\pct_dev\build.xml:239: The following error occurred while executin
g this line:
d:\pcsdev\pct\pct_dev\build.xml:202: Progress procedure failed - No output file

**Comment by axelebnerpcsat** [comment]
When I add my version from pctCompile.p to pct/v10/ then I get the error:
d:\pcsdev\pct\pct_dev\build.xml:303: dlcBin attribute : d:\bin not found

my pct.build.properties settings:
DLC8=\
DLC9=c:\dlc102b
DLC10=\
DLC10-64=\
DLC11=\

Anything else I should do?

**Comment by jakejustus** [comment]
Do you try to recompile PCT ? I think you could just try to use PCT.jar (from the releases page) on your project, that may be easier.
By the way, if you want to recompile PCT, you should remove PCT.jar from $ANT_HOME/lib. There's no reference anymore to pct/v10/pctCompile.p in the code, so I suspect you're using a different version somewhere.

**Comment by axelebnerpcsat** [comment]
Now it took the build version (jenkins-Dev1-PCT-461) as you told me.

When I compile from scratch (using ant -v) with  numThreads=3 then I get 
[PCTCompileExt] 0 file(s) compiled
[PCTCompileExt] Failed to compile 4.799 file(s)
but I get no errors in the logfiles!? 

When I compile from scratch (using ant -v) without  numThreads=3 then I get 
[PCTCompileExt] 4799 file(s) compiled
[PCTCompileExt] Return value : 0

Any ideas?

**Comment by jakejustus** [comment]
Do you use <PCTCompileExt> in your build script ? Or <PCTCompile> with and
without numThreads attribute ?
I'll do the test and verify log messages on my side.

On Sat, Apr 30, 2016 at 10:14 AM, Axel Ebner notifications@github.com
wrote:

> Now it took the build version (jenkins-Dev1-PCT-461) as you told me.
> 
> When I compile from scratch (using ant -v) with numThreads=3 then I get
> [PCTCompileExt] 0 file(s) compiled
> [PCTCompileExt] Failed to compile 4.799 file(s)
> but I get no errors in the logfiles!?
> 
> When I compile from scratch (using ant -v) without numThreads=3 then I get
> [PCTCompileExt] 4799 file(s) compiled
> [PCTCompileExt] Return value : 0
> 
> Any ideas?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-215945951

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by axelebnerpcsat** [comment]
I used PCTCompileExt without numThreads.
My fault - again :(

**Comment by jakejustus** [comment]
No problem :-) Log messages are not part of unit tests, so there might
problems here !

On Sat, Apr 30, 2016 at 10:22 AM, Axel Ebner notifications@github.com
wrote:

> I used without numThreads.
> My fault - again :(
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-215946241

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by axelebnerpcsat** [comment]
The ignoredIncludes works as expected, I'll do some more tests in the evening.
Can the numThreads parameter also used in the PCTCompile?

**Comment by jakejustus** [comment]
Yes, PCTCompileExt is now deprecated (still there for legacy scripts).
numThreads can be set in PCTCompile, which will then decide if a single or
multiple threads compilation have to be started :
https://community.progress.com/community_groups/openedge_development/f/19/t/24665

Gilles

On Sat, Apr 30, 2016 at 10:35 AM, Axel Ebner notifications@github.com
wrote:

> The ignoredIncludes works as expected, I'll do some more tests in the
> evening.
> Can the numThreads parameter also used in the PCTCompile?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-215946804

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by axelebnerpcsat** [comment]
Everything works fine for me :-).

But xrefDir does not work as expected, always writing to [destDir]/.ptc
destDir="..\patidok\pcssw\int"
xrefDir="..\patidok.pct\int"
then all *.crc, *.hierachy, *.inc go to ./patidok/pcssw/int/.pct..

Any hints for me?

Axel

**Comment by jakejustus** [comment]
I'll add a test case for that, as this attribute is not tested. But found a
few other issues in the mean time, so I'll fix them all at once.

On Sun, May 1, 2016 at 12:28 PM, Axel Ebner notifications@github.com
wrote:

> Everything works fine for me :-).
> 
> But xrefDir does not work as expected, always writing to [destDir]/.ptc
> Any hints for me?
> 
> Axel
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-216033280

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by jakejustus** [comment]
Fixed in dev, will update the release page

**Comment by axelebnerpcsat** [comment]
Perfect! Thank you.

**Comment by axelebnerpcsat** [comment]
I only see the v204-pre on https://github.com/jakejustus/pct/releases when can I test the new version?

**Comment by jakejustus** [comment]
Just updated the v204-pre release, you can download again PCT.jar

On Tue, May 3, 2016 at 7:28 PM, Axel Ebner notifications@github.com wrote:

> I only see the v204-pre on https://github.com/jakejustus/pct/releases
> when can I test the new version?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-216604239

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by axelebnerpcsat** [comment]
Hi,

now I'm using jenkins-Dev1-PCT-464

ant > compile.log 2> error.log
Like in the previous version of Dev1 I don't get any compile errors to stderr (error.log), they are logged to the stdout (compile.log) with the other informations.

Interfaces are compiled in to the root dir of the propath. I know it's a progress bug. How do you handle it?

**Comment by axelebnerpcsat** [comment]
xrefDir is OK :+1:

**Comment by jakejustus** [comment]
Hi,

When using the single-threaded version of PCTCompile, stderr is never used
as messages are directly written from the Progress session (which doesn't
know about stderr AFAIK). The multi-threaded version pass messages back to
the Java process, which then output error messages on stderr and
info/verbose on stdout.
Regarding your problem with interfaces, could you open another issue ?

On Wed, May 4, 2016 at 9:37 PM, Axel Ebner notifications@github.com wrote:

> Hi,
> 
> now I'm using jenkins-Dev1-PCT-464
> 
> ant > compile.log 2> error.log
> Like in the previous version of Dev1 I don't get any stderr (error.log),
> errors are shown in the stdout (compile.log) with the other informations.
> 
> Interfaces are compiled in to the root dir of the propath. I know it's a
> progress bug. How do you handle it?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> https://github.com/jakejustus/pct/pull/130#issuecomment-216977454

## 

Gilles QUERRET

Riverside Software
15d route de Bellevue • 69440 Mornant • France
Mob : +33 662.525.532

**Comment by axelebnerpcsat** [comment]
Now I'm back agian!

Yes it is the difference by using the singe/multi-threaded compile. Do you see a way to put all errors to the stderr?

I'll open another issue with the interface.

**Comment by jakejustus** [comment]
stdout/stderr : it was done some time ago using a specific logger, however the result wasn't really pleasant, and wasn't working well with CI tools. Are there specific messages you'd like to see ?

Interfaces : answered on the other issue

**Comment by axelebnerpcsat** [comment]
stdout/stderr : that solution works for me
"ant > compile.log 2>&1"
stdout/stderr goes to the compile.log
